File: Examples/Conversations/Obj-C/ConversationsExample/PDPViewController.m
//
//  PDPViewController.m
//  ConversationsExample
//
//  Copyright © 2016 Bazaarvoice. All rights reserved.
//

#import "PDPViewController.h"
#import "StatisticTableViewCell.h"

typedef enum { ReviewStats = 0, QAStats, StatSectionsCount } StatSections;

typedef enum {
  TotalReviewCount = 0,
  AverageOverallRating,
  HelpfulVoteCount,
  NotHelpfulVoteCount,
  RecommendedCount,
  NotRecommendedCount,
  OverallRatingRange,
  ReviewStateRowsCount
} ReviewStateRows;

typedef enum {
  TotalQuestions = 0,
  TotalAnswers,
  AnswerHelpfulVoteCount,
  AnswerNotHelpfulVoteCount,
  QuestionHelpfulVoteCount,
  QuestionNotHelpfulVoteCount,
  QAStatRowsCount
} QAStatRows;

@interface PDPViewController () <UITableViewDataSource>

@property(weak, nonatomic) IBOutlet UITableView *demoStatsTableView;

@property BVReviewStatistics *reviewStatistics;
@property BVQAStatistics *questionAnswerStats;

@end

@implementation PDPViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  self.demoStatsTableView.estimatedRowHeight = 80;
  self.demoStatsTableView.rowHeight = UITableViewAutomaticDimension;
  [self.demoStatsTableView
                 registerNib:[UINib nibWithNibName:@"StatisticTableViewCell"
                                            bundle:nil]
      forCellReuseIdentifier:@"StatisticTableViewCell"];

  BVProductDisplayPageRequest *request =
      [[BVProductDisplayPageRequest alloc] initWithProductId:@"test1"];
  [request includeStatistics:PDPContentTypeReviews];
  [request includeStatistics:PDPContentTypeAnswers];
  [request includeStatistics:PDPContentTypeQuestions];
  [request includeContent:PDPContentTypeQuestions limit:10];
  [request sortIncludedQuestions:BVSortOptionQuestionsTotalAnswerCount
                           order:BVSortOrderDescending];

  [request load:^(BVProductsResponse *_Nonnull response) {

    self.reviewStatistics = response.result.reviewStatistics;
    self.questionAnswerStats = response.result.qaStatistics;
    [self.demoStatsTableView reloadData];

  }
      failure:^(NSArray<NSError *> *_Nonnull errors) {
        NSLog(@"ERROR: %@", errors.description);
      }];
}

#pragma mark UITableViewDatasource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return StatSectionsCount;
}

- (NSString *)tableView:(UITableView *)tableView
    titleForHeaderInSection:(NSInteger)section {
  switch (section) {
  case ReviewStats:
    return @"Product Review Statistics";
    break;
  case QAStats:
    return @"Product Question & Answer Statistics";
    break;

  default:
    break;
  }

  return @"";
}

- (NSInteger)tableView:(UITableView *)tableView
    numberOfRowsInSection:(NSInteger)section {

  switch (section) {
  case ReviewStats:
    return ReviewStateRowsCount;
    break;
  case QAStats:
    return QAStatRowsCount;
    break;
  default:
    break;
  }

  return 0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {

  StatisticTableViewCell *cell =
      [tableView dequeueReusableCellWithIdentifier:@"StatisticTableViewCell"];

  switch (indexPath.section) {
  case ReviewStats:

    // Review stats
    switch (indexPath.row) {
    case TotalReviewCount:
      cell.statTypeLabel.text = @"Total Review Count";
      cell.statValueLabel.text =
          [self.reviewStatistics.totalReviewCount stringValue];
      break;

    case AverageOverallRating:
      cell.statTypeLabel.text = @"Average Overall Rating";
      cell.statValueLabel.text =
          [self.reviewStatistics.averageOverallRating stringValue];
      break;

    case HelpfulVoteCount:
      cell.statTypeLabel.text = @"Helpful Vote Count";
      cell.statValueLabel.text =
          [self.reviewStatistics.helpfulVoteCount stringValue];
      break;

    case NotHelpfulVoteCount:
      cell.statTypeLabel.text = @"Not Helpful Vote Count";
      cell.statValueLabel.text =
          [self.reviewStatistics.notHelpfulVoteCount stringValue];
      break;

    case RecommendedCount:
      cell.statTypeLabel.text = @"Recommended Count";
      cell.statValueLabel.text =
          [self.reviewStatistics.recommendedCount stringValue];
      break;

    case NotRecommendedCount:
      cell.statTypeLabel.text = @"Not Recommended Count";
      cell.statValueLabel.text =
          [self.reviewStatistics.notRecommendedCount stringValue];
      break;

    case OverallRatingRange:
      cell.statTypeLabel.text = @"Overall Rating Range";
      cell.statValueLabel.text =
          [self.reviewStatistics.overallRatingRange stringValue];
      break;

    default:
      break;
    }

    break;

  case QAStats:

    switch (indexPath.row) {
    case TotalQuestions:
      cell.statTypeLabel.text = @"Total Questions";
      cell.statValueLabel.text =
          [self.questionAnswerStats.totalQuestionCount stringValue];
      break;

    case TotalAnswers:
      cell.statTypeLabel.text = @"Total Answers";
      cell.statValueLabel.text =
          [self.questionAnswerStats.totalAnswerCount stringValue];
      break;

    case AnswerHelpfulVoteCount:
      cell.statTypeLabel.text = @"Answer Helpful Vote Count";
      cell.statValueLabel.text =
          [self.questionAnswerStats.answerHelpfulVoteCount stringValue];
      break;

    case AnswerNotHelpfulVoteCount:
      cell.statTypeLabel.text = @"Answer Not Helpful Vote Count";
      cell.statValueLabel.text =
          [self.questionAnswerStats.answerNotHelpfulVoteCount stringValue];
      break;

    case QuestionHelpfulVoteCount:
      cell.statTypeLabel.text = @"Question Helpful Vote Count";
      cell.statValueLabel.text =
          [self.questionAnswerStats.questionHelpfulVoteCount stringValue];
      break;

    case QuestionNotHelpfulVoteCount:
      cell.statTypeLabel.text = @"Question Not Helpful Vote Count";
      cell.statValueLabel.text =
          [self.questionAnswerStats.questionNotHelpfulVoteCount stringValue];
      break;

    default:
      break;
    }

    break;

  default:
    break;
  }

  return cell;
}

@end

File: Examples/Conversations/Swift/ConversationsExample/PDPViewController.swift
//
//  PDPViewController.swift
//  Conversations
//
//  Copyright © 2016 Bazaarvoice. All rights reserved.
//

import UIKit
import BVSDK

class PDPViewController: BVProductDisplayPageViewController, UITableViewDataSource {
    
    @IBOutlet weak var demoStatsTableView: UITableView!
    
    enum StatSections : Int {
        case ReviewStats
        case QAStats
        
        static var count: Int { return StatSections.QAStats.hashValue + 1}
    }
    
    enum ReviewStateRows : Int {
        case TotalReviewCount
        case AverageOverallRating
        case HelpfulVoteCount
        case NotHelpfulVoteCount
        case RecommendedCount
        case NotRecommendedCount
        case OverallRatingRange
        
        static var count: Int { return ReviewStateRows.OverallRatingRange.hashValue + 1}
    }
    
    enum QAStatRows : Int {
        case TotalQuestions
        case TotalAnswers
        case AnswerHelpfulVoteCount
        case AnswerNotHelpfulVoteCount
        case QuestionHelpfulVoteCount
        case QuestionNotHelpfulVoteCount
        
        static var count: Int { return QAStatRows.QuestionNotHelpfulVoteCount.hashValue + 1}
    }
    
    var reviewStats = BVReviewStatistics()
    var questionAnswerStats = BVQAStatistics()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        demoStatsTableView.estimatedRowHeight = 80
        demoStatsTableView.rowHeight = UITableViewAutomaticDimension
        demoStatsTableView.registerNib(UINib(nibName: "StatisticTableViewCell", bundle: nil), forCellReuseIdentifier: "StatisticTableViewCell")
        
        let productPage = BVProductDisplayPageRequest(productId: "test1")
        .includeStatistics(.Reviews)
        .includeStatistics(.Questions)
        .includeStatistics(.Answers)
        
        productPage.load({ (response) in
            
            self.reviewStats         = (response.result?.reviewStatistics)!
            self.questionAnswerStats = (response.result?.qaStatistics)!
            
            self.demoStatsTableView.reloadData()
            
            }) { (error) in
                print(error)
        }
        
    }
    
    // MARK: UITableViewDatasource
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return StatSections.count
    }
    
    func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        
        switch section {
            case StatSections.ReviewStats.rawValue:
                return "Product Review Statistics"
            case StatSections.QAStats.rawValue:
                return "Product Question & Answer Statistics"
            default:
            return ""
        }
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
        case StatSections.ReviewStats.rawValue:
            return ReviewStateRows.count
        case StatSections.QAStats.rawValue:
            return QAStatRows.count
        default:
            return 0
        }
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCellWithIdentifier("StatisticTableViewCell")! as! StatisticTableViewCell
        
        switch indexPath.section {
            
            case StatSections.ReviewStats.rawValue:
                        
            switch indexPath.row {
                case ReviewStateRows.TotalReviewCount.rawValue:
                    cell.statTypeLabel.text = "Total Review Count"
                    cell.statValueLabel.text = reviewStats.totalReviewCount?.stringValue
                case ReviewStateRows.AverageOverallRating.rawValue:
                    cell.statTypeLabel.text = "Average Overall Rating"
                    cell.statValueLabel.text = reviewStats.averageOverallRating?.stringValue
                case ReviewStateRows.HelpfulVoteCount.rawValue:
                    cell.statTypeLabel.text = "Helpful Vote Count"
                    cell.statValueLabel.text = reviewStats.helpfulVoteCount?.stringValue
                case ReviewStateRows.NotHelpfulVoteCount.rawValue:
                    cell.statTypeLabel.text = "Not Helpful Vote Count"
                    cell.statValueLabel.text = reviewStats.notHelpfulVoteCount?.stringValue
                case ReviewStateRows.RecommendedCount.rawValue:
                    cell.statTypeLabel.text = "Recommended Count"
                    cell.statValueLabel.text = reviewStats.recommendedCount?.stringValue
                case ReviewStateRows.NotRecommendedCount.rawValue:
                    cell.statTypeLabel.text = "Not Recommeded Count"
                    cell.statValueLabel.text = reviewStats.notRecommendedCount?.stringValue
                case ReviewStateRows.OverallRatingRange.rawValue:
                    cell.statTypeLabel.text = "Overall Rating Range"
                    cell.statValueLabel.text = reviewStats.overallRatingRange?.stringValue
                default:
                    cell.statTypeLabel.text = "Error"
                    cell.statValueLabel.text = "Error"

            }
            
        case StatSections.QAStats.rawValue:
            
            switch indexPath.row {
                case QAStatRows.TotalQuestions.rawValue:
                    cell.statTypeLabel.text = "Total Question Count"
                    cell.statValueLabel.text = questionAnswerStats.totalQuestionCount?.stringValue
                case QAStatRows.TotalAnswers.rawValue:
                    cell.statTypeLabel.text = "Total Answer Count"
                    cell.statValueLabel.text = questionAnswerStats.totalAnswerCount?.stringValue
                case QAStatRows.AnswerHelpfulVoteCount.rawValue:
                    cell.statTypeLabel.text = "Answer Helpful Vote Count"
                    cell.statValueLabel.text = questionAnswerStats.answerHelpfulVoteCount?.stringValue
                case QAStatRows.AnswerNotHelpfulVoteCount.rawValue:
                    cell.statTypeLabel.text = "Answer Not Helpful Vote Count"
                    cell.statValueLabel.text = questionAnswerStats.answerNotHelpfulVoteCount?.stringValue
                case QAStatRows.QuestionHelpfulVoteCount.rawValue:
                    cell.statTypeLabel.text = "Question Helpful Vote Count"
                    cell.statValueLabel.text = questionAnswerStats.questionHelpfulVoteCount?.stringValue
                case QAStatRows.QuestionNotHelpfulVoteCount.rawValue:
                    cell.statTypeLabel.text = "Question Not Helpful Vote Count"
                    cell.statValueLabel.text = questionAnswerStats.questionNotHelpfulVoteCount?.stringValue
                default:
                    cell.statTypeLabel.text = "Error"
                    cell.statValueLabel.text = "Error"
            }
            
        default:
            print("Bad section provided for stats table view")
        }
        
        return cell
    }
    
}
